src/Schemes/AffineSchemes/Objects/Attributes.jl

```diff
@@ -389,14 +396,246 @@ function set_name!(X::AbsSpec, name::String)
+
+```
+"""
+@attr function reduced_scheme(X::AbsSpec{<:Field, <:MPolyQuoLocalizedRing})
+  I = modulus(OO(X))
+  J = radical(pre_saturated_ideal(I))
+  return Spec(base_ring(J), J, inverted_set(OO(X)))
+end
+
+@attr function reduced_scheme(X::AbsSpec{<:Field, <:MPolyQuo})
+  J = radical(modulus(OO(X)))
+  return Spec(base_ring(J), J)
+end
+
+## to make reduced_scheme agnostic for quotient ring
+@attr AbsSpec function reduced_scheme(X::AbsSpec{<:Ring, <:MPAnyNonQuoRing})
+  return X
+end
+
+### TODO: The following two functions (singular_locus, 
+###       singular_locus_reduced) need to be made type-sensitive 
+###       and reduced=true needs to be set automatically for varieties 
+###       as soon as not only schemes, but also varieties as separte
+###       type have been introduced in OSCAR
+### TODO: Make singular locus also available for projective schemes and
+###       for covered schemes (using the workhorse here...).
+ 
+@doc Markdown.doc"""
+    singular_locus(X::Scheme{<:Field}) -> Scheme
+
+Return the singular locus of `X`.
+
+For computing the singular locus of the reduced scheme induced by `X`,
+please use [`singular_locus_reduced`](@ref).
+
+Currently this command is available for affine schemes and for space germs.
+TODO: Covered schemes, projective schemes
+
+Over non-perfect fields, this command returns the non-smooth locus and `X`
+may still be regular at some points of the returned subscheme.
+
+See also [`is_smooth`](@ref).
+
+# Examples
+``` jldoctest
+julia> R, (x,y,z) = QQ["x", "y", "z"]
+(Multivariate Polynomial Ring in x, y, z over Rational Field, fmpq_mpoly[x, y, z])
+
+julia> I = ideal(R, [x^2 - y^2 + z^2])
+ideal(x^2 - y^2 + z^2)
+
+julia> A3 = Spec(R)
+Spec of Multivariate Polynomial Ring in x, y, z over Rational Field
+
+julia> X = Spec(R,I)
+Spec of Quotient of Multivariate Polynomial Ring in x, y, z over Rational Field by ideal(x^2 - y^2 + z^2)
+
+julia> singular_locus(A3)
+Spec of Quotient of Multivariate Polynomial Ring in x, y, z over Rational Field by ideal(1)
 
+julia> singular_locus(X)
+Spec of Quotient of Multivariate Polynomial Ring in x, y, z over Rational Field by ideal(x^2 - y^2 + z^2, 2*x, -2*y, 2*z)
 
+julia> U = MPolyComplementOfKPointIdeal(R,[0,0,0])
+complement of maximal ideal corresponding to point with coordinates fmpq[0, 0, 0]
+
+julia> Y = Spec(R,I,U)
+Spec of Localization of Quotient of Multivariate Polynomial Ring in x, y, z over Rational Field by ideal(x^2 - y^2 + z^2) at the multiplicative set complement of maximal ideal corresponding to point with coordinates fmpq[0, 0, 0]
+
+julia> singular_locus(Y)
+Spec of Localization of Quotient of Multivariate Polynomial Ring in x, y, z over Rational Field by ideal(x^2 - y^2 + z^2, x^2 - y^2 + z^2, 2*x, -2*y, 2*z, x^2 - y^2 + z^2, x^2 - y^2 + z^2) at the multiplicative set complement of maximal ideal corresponding to point with coordinates fmpq[0, 0, 0]
+
+```
+"""
+function singular_locus(X::AbsSpec{<:Field, <:MPAnyQuoRing})
```

Comment:
Caching the singular locus does not make sense in general, as it can easily become really large (ideal of minors) and is not ubiquitous in applications. Inferring is_smooth on the other hand, is worth it and I added it to the code..